### PR TITLE
Scale ratelimit with ingress pods

### PR DESCRIPTION
**Pull request by dbaumgarten**
### Proposed changes

Implements #4894

Add a flag (in VS-Policy and in Ingress-Annotations) to enable ratelimit-scaling. If enabled, the configured ratelimit is divided by the number of currently active ingress-pods. This way the total sum of the ratelimit stays constant, even when autoscaling changes the number of nginx-ingress pods.

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by ADubhlaoich — COMMENTED** [review]
LGTM docs-wise, but most of the PR is code-based, so I am not approving.

**Comment by dbaumgarten** [comment]
@shaun-nx What do you think? Does this look fine or are any changes required?

**Comment by brianehlert** [comment]
Just an FYI as this has been lingering for review.
The NGINX Plus version of the implementation should get accurate rate limiting across all pods with state sync.
That comes with the shared memory zone being configured for the rate limit and then state sharing being enabled.

File: docs/content/configuration/ingress-resources/advanced-configuration-with-annotations.md [hunk 1]
```diff
@@ -186,6 +186,7 @@ The table below summarizes the available annotations.
 |``nginx.org/limit-req-dry-run`` | N/A | Enables the dry run mode. In this mode, the rate limit is not actually applied, but the number of excessive requests is accounted as usual in the shared memory zone. | false | true |
 |``nginx.org/limit-req-log-level`` | N/A | Sets the desired logging level for cases when the server refuses to process requests due to rate exceeding, or delays request processing. Allowed values are info, notice, warn or error. | error | info |
 |``nginx.org/limit-req-reject-code`` | N/A | Sets the status code to return in response to rejected requests. Must fall into the range 400..599. | 429 | 503 |
+|``nginx.org/limit-req-scale`` | N/A | If set to true, the configured rate is divided by the number of nginx-ingress pods currently serving traffic. This maintains a constant rate-limit, even if the number of nginx-pods changes due to autoscaling. | false | true |
```
Comment (j1m-ryan) [inline docs/content/configuration/ingress-resources/advanced-configuration-with-annotations.md:189]:
```suggestion
|``nginx.org/limit-req-scale`` | N/A | Enables a constant rate-limit by dividing the configured rate by the number of nginx-ingress pods currently serving traffic. This adjustment ensures that the rate-limit remains consistent, even as the number of nginx-pods fluctuates due to autoscaling. | false | true |
```

File: docs/content/configuration/policy-resource.md [hunk 2]
```diff
@@ -124,6 +124,7 @@ rateLimit:
 |``dryRun`` | Enables the dry run mode. In this mode, the rate limit is not actually applied, but the number of excessive requests is accounted as usual in the shared memory zone. | ``bool`` | No |
 |``logLevel`` | Sets the desired logging level for cases when the server refuses to process requests due to rate exceeding, or delays request processing. Allowed values are ``info``, ``notice``, ``warn`` or ``error``. Default is ``error``. | ``string`` | No |
 |``rejectCode`` | Sets the status code to return in response to rejected requests. Must fall into the range ``400..599``. Default is ``503``. | ``int`` | No |
+|``scale`` | When set to true, the configured ratelimit is divided by the number of currently active ingress-pods. This ensures a consistent ratelimit, even when autoscaling is enabled. | ``bool`` | No |
```
Comment (j1m-ryan) [inline docs/content/configuration/policy-resource.md:127]:
```suggestion
|``scale`` | Enables a constant rate-limit by dividing the configured rate by the number of nginx-ingress pods currently serving traffic. This adjustment ensures that the rate-limit remains consistent, even as the number of nginx-pods fluctuates due to autoscaling. | ``bool`` | No |
```

**Comment by dbaumgarten** [comment]
Rebased on current master and added a commit clarifying the docs about the limitations of this feature.

**Comment by j1m-ryan** [comment]
Hi @dbaumgarten, thanks for following through, just an FYI that github actions are not working properly at the moment, that is why the tests are not running. https://www.githubstatus.com/

**Review by j1m-ryan — APPROVED** [review]
lgtm 🚀

**Review by pdabelf5 — APPROVED** [review]

File: internal/k8s/controller.go [hunk 3]
```diff
@@ -910,6 +915,60 @@ func (lbc *LoadBalancerController) syncEndpointSlices(task task) bool {
...
+		}
+		for _, ingress := range resourceExes.MergeableIngresses {
+			found = true
+			_, err := lbc.configurator.AddOrUpdateMergeableIngress(ingress)
+			if err != nil {
+				glog.Errorf("Error updating ratelimit for Ingress %s/%s: %s", ingress.Master.Ingress.Namespace, ingress.Master.Ingress.Name, err)
+			}
+		}
+
+		// handle virtualservers
```
Comment (vepatel) [inline internal/k8s/controller.go:946]: @dbaumgarten something we observed in regression, the code below was causing panic when CRD flag was not set so we've wrapped it in

```
if lbc.areCustomResourcesEnabled {}
```

Comment (dbaumgarten) [reply internal/k8s/controller.go:946]: Oh. Nice find.
Yes, wrapping it in that if sounds like a viable solution 👍

**Pull request merged**